Add initial Earth, Moon, Satellite animation example to v2

karthikeyan4747 opened the pull request:
## Summary

This PR adds an initial v2 version of the **Earth, Moon, and Satellite animation example** by porting the classic “Texture Sphere Animation” tutorial from the `master` branch into the `v2` branch.

The goal is to recreate the same visual/animation behavior but using the v2 structure, so we can refine it together.

---

## What’s Included

- Added new example file:  
  **`docs/examples/viz_earth_animation.py`**
- Ported all logic from the original tutorial:
  - Textured Earth sphere
  - Textured Moon sphere
  - Marker + label on Earth
  - Satellite model orbiting the Moon
  - Camera transitions and zoom sequences
  - Timer-based animation via `ShowManager`
- Followed the style/pattern used in other v2 example files.
- Wrapped the example in a `main(interactive=True)` function to match v2 convention.

---

## Current Issue (Need Guidance)

`texture_on_sphere` is not available in the v2 branch:


Since the original tutorial relies on this helper for textured spheres, I kept the calls as placeholders so we can decide:

- whether v2 has a new WebGPU-compatible method for textured spheres,
- or whether this helper should be reintroduced/exposed in the v2 API,
- or whether a different actor creation pipeline should be used.

Happy to update this once I get direction.

---

## Questions

1. Is `docs/examples/viz_earth_animation.py` the correct location in the v2 example layout?
2. What is the recommended v2/WebGPU API to replace `texture_on_sphere`?
3. Should example textures/models be fetched using a different mechanism in v2?

---

## Notes

This is a **draft PR** so we can iterate.  
Once the correct v2 API for textured spheres is confirmed, I’ll update the example accordingly.

karthikeyan4747 committed:
Add initial Earth, Moon, Satellite animation example to v2

maharshi-gor commented:
Hey @karthikeyan4747 ,

Thanks for creating this.

Answer to you questions.

1. The location is correct, but you are required to add this example in the valid_examples.toml file which will pick this example while generating the documentation.

2. We currently do not have direct API to provide texture_on_sphere. We can create it using surface actor in which we will provide the sphere and texture. If you would create an API in the actor/curved.py that will be amazing.

3. No, textures are fetched exactly same in the v2, at least from usage of API signature.

Now, general advise.

- We do not use interactive variables any more if you see other tutorials in v2. System has an environment variable called FURY_OFFSCREEN which decides on the interactivity.
- Provide more information of the code similar to what other tutorials have as it is for beginners to try FURY.

Again, thanks for the contribution!

karthikeyan4747 committed:
Added the Earth animation to Valid examples

karthikeyan4747 committed:
Add texture_on_sphere API and minimal Earth texture example for v2

karthikeyan4747 commented:
Hey @maharshi-gor,

I’ve pushed the minimal Earth example along with the texture_on_sphere API.
I tried really hard to port the full animation from the master branch — went through the v2 codebase, experimented with different approaches, and even used AI to understand missing parts — but a lot of the older functions (rotations, camera transforms, timers, etc.) aren’t available or work differently in v2.

To be completely honest, I feel a bit lost in this sea of code right now. I’m still learning how to navigate large projects like this, and I don’t want to force hacky or unreliable code. So I pushed the cleanest minimal version that I could get running consistently.

Also, on my local machine I keep hitting a GLFW initialization error before the window opens, so I couldn’t visually inspect the final rendering — but the code runs up to that point.

I really want to improve and continue contributing (I’m aiming for internships, GSoC, and real open-source experience), so if you can point me toward the correct direction for v2 transforms/animation, I’d be glad to keep building on this.

Thanks again for your patience and guidance.
— Karthi

maharshi-gor commented:
Hey @karthikeyan4747 `texture_on_sphere` function is at the wrong place you need to have proper settings in the editor.

Please configure your editor to show errors for python. 

Take your time to look at the codebase and you will be able to navigate. Let's just start with rendering the object first.

karthikeyan4747 committed:
Fixed the loaction of Texture_on_sphere and Implemented changes

karthikeyan4747 commented:
Hey @maharshi-gor,

I’ve updated the PR with a clean and stable Earth texture example that works with the current v2 branch. After exploring the codebase more deeply, I realized many of the animation- and camera-related APIs used in the master tutorial (set_camera, rotation utilities, timer callbacks, transforms, etc.) are not yet available in v2, so I kept this example static and fully compatible.

I also made sure to place the new texture_on_sphere API in the correct module (actor/curved.py) and added clear comments and explanations throughout the code, as you suggested. The implementation now uses the v2 surface actor internally and produces correct UV-mapped textured spheres.

One thing I wanted to mention: on my system the WebGPU backend throws a shader validation error (GPUValidationError coming from wgpu-native) when rendering begins. The script runs fine up to that point, but the window displays a black screen. Since the error originates inside the WebGPU shader pipeline, I assume it’s environment/GPU related rather than an issue with the example itself. Just mentioning it in case it helps with testing or future debugging.

Once the v2 animation / transform / camera APIs are finalized, I’d be happy to extend this example further and bring back more of the original behavior. For now, this version focuses on correctness, clarity, and stability.

Thanks again for guiding me through this — still learning how to navigate larger projects, but this has already taught me a lot.

— Karthi

maharshi-gor commented:
Hello Karthi,

You are required to register the function in the actor/__init__.pyi for it to be used.

maharshi-gor commented:
I tried adding it to that. But I see the seam at joints. Please find it in the screenshot attached.

Let's check why we are getting that seam.
<img width="539" height="562" alt="Screenshot 2025-12-12 at 4 18 49 PM" src="https://github.com/user-attachments/assets/f065a11d-0f93-48f8-9fcb-98d9a7de056e" />

karthikeyan4747 commented:
Yeah, I see the seam.

From what I understand, this usually happens at the texture join on a sphere (where the texture starts and ends). It could be because the texture wrapping on the U direction isn’t set to repeat, or the earth texture itself has a small mismatch at the edges, which becomes visible.

I dont really have any other clue why this occurs

maharshi-gor commented:
Would you like to take up this and try fixing if you want to explore. I think this will be a great help. 

Try looking for an answer and understand the coordinates wrapping. This might not come directly as the answer in the books you might need to understand which will help you debug and understand issues on the repo and help you provide knowledge to debug the code.

karthikeyan4747 commented:
Yeah, I’ll take this up.
I’ll dig into the coordinate wrapping, understand where it breaks, and fix it properly.
If it’s not in the books, that’s fine — I’ll reason it out from the code and the math.
I’ll need a little time to understand the code properly since I also have some college work, but I’ll debug it and come back with a clear fix and explanation within a week or two.

maharshi-gor commented:
Hello @karthikeyan4747 any update on this?

karthikeyan4747 commented:
Hello @maharshi-gor really sorry for the late reply as I had my semester exams for the past few weeks I wasn't able to respond properly I'll start working on the issue and update you once I find a answer

ganimtron-10 commented:
Hi @karthikeyan4747 ,

It's been a while since the last update on this PR. Could you let us know if you're still working on this? If we don't see any progress by next week, I’ll go ahead and close it to keep the queue manageable.

If you need any help or clarification to move forward, just let us know. Thanks!

karthikeyan4747 commented:
Hey @ganimtron-10, thanks for the nudge! I've shifted my focus to a different project for now and won't be able to continue with this PR. Feel free to close it. Appreciate the patience and the work you're doing on the project! 🙏
